src/main/java/seedu/mark/logic/commands/ImportCommand.java

```diff
@@ -0,0 +1,240 @@
+            throw new CommandException(String.format(MESSAGE_FILE_FORMAT_INCORRECT, filePath));
+        }
+
+        if (newMark.isEmpty()) {
+            throw new CommandException(String.format(MESSAGE_FILE_NOT_FOUND, filePath));
+        }
+
+        return newMark.get();
+    }
+
+    /**
+     * Executes the import command on the given {@code model} and {@code storage}.
+     *
+     * @param model {@code Model} which the command should operate on.
+     * @param storage {@code Storage} which the command should operate on.
+     * @return A CommandResult indicating a successful import, and (if applicable)
+     *         shows bookmarks that were skipped during import.
+     * @throws CommandException if an error occurs while reading bookmarks from Storage
+     */
+    @Override
+    public CommandResult execute(Model model, Storage storage) throws CommandException {
+        requireAllNonNull(model, storage);
+
+        ReadOnlyMark newMark = readMarkFromStorage(storage, filePath);
+
+        MarkImporter importer = new MarkImporter(model, newMark);
+        importer.importFolders();
+
+        if (!importer.hasBookmarksToImport()) {
+            return new CommandResult(String.format(MESSAGE_NO_BOOKMARKS_TO_IMPORT,
+                    importer.getExistingBookmarksAsString()));
+        }
+        importer.importBookmarks();
+
+        // TODO: rewrite messages to indicate whether folders were imported
+        String message = importer.hasExistingBookmarks()
+                ? String.format(MESSAGE_IMPORT_SUCCESS_WITH_DUPLICATES, filePath,
+                    importer.getExistingBookmarksAsString())
+                : String.format(MESSAGE_IMPORT_SUCCESS, filePath);
+        return new CommandResult(message);
+    }
+
+    /**
+     * Checks whether an object is equal to this {@code ImportCommand}.
+     *
+     * @param other Object to compare.
+     * @return true if the object is an ImportCommand with the same {@code filePath}
+     * as the current {@code ImportCommand}.
+     */
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof ImportCommand // instanceof handles nulls
+                && filePath.equals(((ImportCommand) other).filePath)); // state check
+    }
+
+    /**
+     * Stores data from a {@code Mark} to import them into a {@code Model}.
+     */
+    public static class MarkImporter {
+        private Model model;
+        private FolderStructure foldersToImport;
+        private List<Bookmark> existingBookmarks = new ArrayList<>();
+        private List<Bookmark> bookmarksToImport = new ArrayList<>();
+
+        MarkImporter(Model model, ReadOnlyMark markToImport) {
+            this.model = model;
+            this.foldersToImport = markToImport.getFolderStructure();
+            for (Bookmark bookmark : markToImport.getBookmarkList()) {
```

Comment:
Executing logic in the constructor seems a bit weird, but I'm not sure how to fix it without making `ImportCommand` call another `MarkImporter` method after instantiating it.